pallets/loans/src/entities/loans.rs

```diff
@@ -321,26 +311,46 @@ impl<T: Config> ActiveLoan<T> {
 		self.total_borrowed.ensure_add_assign(amount)?;
 
 		match &mut self.pricing {
-			ActivePricing::Internal(inner) => inner.adjust_debt(Adjustment::Increase(amount))?,
-			ActivePricing::External(inner) => inner.adjust_debt(Adjustment::Increase(amount))?,
+			ActivePricing::Internal(inner) => inner.adjust(Adjustment::Increase(amount))?,
+			ActivePricing::External(inner) => {
+				inner.adjust(Adjustment::Increase(amount), Zero::zero())?
+			}
 		}
 
 		Ok(())
 	}
 
-	fn ensure_can_repay(&self, amount: T::Balance) -> Result<T::Balance, DispatchError> {
-		let max_repay_amount = match &self.pricing {
-			ActivePricing::Internal(inner) => inner.calculate_debt()?,
-			ActivePricing::External(inner) => inner.calculate_debt()?,
+	fn ensure_can_repay(
+		&self,
+		mut amount: RepaidAmount<T::Balance>,
+	) -> Result<RepaidAmount<T::Balance>, DispatchError> {
+		let (interest_accrued, max_repay_principal) = match &self.pricing {
+			ActivePricing::Internal(inner) => {
+				let principal = self
+					.total_borrowed
+					.ensure_sub(self.total_repaid.principal)?;
+
+				(inner.current_interest(principal)?, principal)
+			}
+			ActivePricing::External(inner) => {
+				(inner.current_interest()?, inner.outstanding_amount()?)
+			}
 		};
 
-		let amount = amount.min(max_repay_amount);
+		amount.interest = amount.interest.min(interest_accrued);
+
+		ensure!(
+			amount.principal <= max_repay_principal,
+			Error::<T>::from(RepayLoanError::MaxPrincipalAmountExceeded)
+		);
 
 		ensure!(
 			match self.restrictions.repayments {
 				RepayRestrictions::None => true,
 				RepayRestrictions::FullOnce => {
-					self.total_repaid.is_zero() && amount == max_repay_amount
+					self.total_repaid.effective()?.is_zero()
```

Comment:
Can we change the `RepayRestrictions` for an active loan? I yes, this `is_zero()` check will prevent repaying the loan. I also think it is not adding any value or what should the problem be, with having a non zero value here?

Reply:
We can not change repay restrictions right now (not sure if we should in the future).

`FullOnce` is, in fact, two restrictions (that in the future will be split: #1353). We want to repay the entire amount, and we want to repay it once. 

But thinking more on it, I think, as you say, checking `is_zero()` makes no sense. And even could generate inconsistency. i.e. You borrow again, but you can not repay again because you repaid the first time. Probably the correct restriction is just `Full`. Good catch!